Added GQIDMS tutorial (Kata) (#3213)
SkylineCommunications/dataminer-docs: user-guide/Advanced_Modules/Dashboards_and_Low_Code_Apps/GQI/Extensions/Ad_hoc_data/Ad_hoc_Tutorials_GQIDMS.md

---
uid: Ad_hoc_Tutorials_GQIDMS
---

# Building a GQI data source that retrieves data from a DMS

In this tutorial, you will learn how you can create a GQI data source that retrieves the active client connections from your DMS.

Expected duration: 15 minutes.

> [!NOTE]
> This tutorial uses DataMiner version 10.3.4/10.4.0 onwards.

## Prerequisites

- [Visual Studio](https://visualstudio.microsoft.com/downloads/) with [DataMiner Integration Studio](xref:Installing_and_configuring_DataMiner_Integration_Studio)

## Overview

- [Step 1: Create the ad hoc data source](#step-1-create-the-ad-hoc-data-source)
- [Step 2: Fetch the client connections from the DMS](#step-2-fetch-the-client-connections-from-the-dms)
- [Step 3: Parse the responses into a GQIPage](#step-3-parse-the-responses-into-a-gqipage)

## Step 1: Create the ad hoc data source

1. Create a new class that implements the [IGQIDatasource](xref:GQI_IGQIDataSource) interface.

> > [!NOTE]

Reviewer: I think this note isn't nested correctly. In the preview, the numbered list now goes 1, 1, 2

> If certain types cannot be found in the file, verify if the *Skyline.DataMiner.Dev.Automation* NuGet package has the correct version. Go to *Tools* > *NuGet Package Manager* > *Manage NuGet Packages for Solution*. Select *Skyline.DataMiner.Dev.Automation*, and verify whether the version installed for the current project is at least *10.3.4*.

> 1. Implement the GetColumns function to define the columns that the data source will contain. Our data source will include two columns: `Client` (of type `String`) and `Connection Time` (of type `DateTime`).

Reviewer: In C# member functions in classes and interfaces are always called 'methods' instead. Also use consistent formatting for code elements vs. names.

Could also link each method reference to its paragraph in the docs.

```suggestion
1. Implement the `GetColumns` method to define the columns of the data source. Our data source will include two columns: a *Client* column of  type `String` and a *Connection Time* column of type `DateTime`.
```

1. Implement the `GetNextPage` function to provide the actual data. For now, leave it empty since we still need to retrieve the data from the DMS.

```csharp

> [GQIMetaData(Name = "Client connections")]

Reviewer: This GQIMetaData isn't mentioned. Should we mention it, or remove it from the code example?

public class ClientConnections : IGQIDataSource
{
    public GQIColumn[] GetColumns()
    {
        return new GQIColumn[]
        {
            new GQIStringColumn("Client"),
            new GQIDateTimeColumn("Connection Time"),
        };
    }

    public GQIPage GetNextPage(GetNextPageInputArgs args)
    {
        // TODO: Fetch the data
        return default;
    }
}
```

## Step 2: Fetch the client connections from the DMS

> To retrieve the current client connections, communication with the DMS is required. You can use the [GQIDMS](xref:GQI_GQIDMS) object, available in the [OnInitInputArgs](xref:GQI_OnInitInputArgs) argument of the `OnInit` function, by implementing the [IGQIOnInit](xref:GQI_IGQIOnInit) interface.

Reviewer: ```suggestion
To retrieve the current client connections, the data source will have to communicate with the DMS. You can use the [GQIDMS](xref:GQI_GQIDMS) object for this. It is available in the [OnInitInputArgs](xref:GQI_OnInitInputArgs) argument of the `OnInit` method by implementing the [IGQIOnInit](xref:GQI_IGQIOnInit) interface.
```

> To retrieve the current client connections, communication with the DMS is required. You can use the [GQIDMS](xref:GQI_GQIDMS) object, available in the [OnInitInputArgs](xref:GQI_OnInitInputArgs) argument of the `OnInit` function, by implementing the [IGQIOnInit](xref:GQI_IGQIOnInit) interface.

Reviewer: Also, maybe we should link here to "Example of retrieving data by means of DMS messages" directly?

> 1. Add the [IGQIOnInit](xref:GQI_IGQIOnInit) interface to the class.

Reviewer: Code blocks are probably not indented correctly. The steps here go 1, 2, 1 now

> 1. Implement the `OnInit` function, which receives [OnInitInputArgs](xref:GQI_OnInitInputArgs) as input arguments. Store the DMS property from the arguments in the class for later use.

Reviewer: We should call regular "input arguments" to C# methods "parameters".
Also to avoid confusion with the actual data source input arguments.

Reviewer: This step should probably be split up? (see doc guidelines about avoiding steps with multiple actions)

```csharp
[GQIMetaData(Name = "Client connections")]
public class ClientConnections : IGQIDataSource
{
    private GQIDMS _dms;

    public OnInitOutputArgs OnInit(OnInitInputArgs args)
    {
        _dms = args.DMS;
        return default;
    }
}
```

> 1. We can use the `GQIDMS` object to send messages to the DMS, to retrieve the client connections we will need the `GetInfoMessage`. The response to that message are multiple `LoginInfoResponseMessage` objects.

Reviewer: Avoid "we" and try make steps are actionable. For example:

```suggestion
1. Use the `GQIDMS` object to send messages to the DMS. Retrieve the client connections by sending a `GetInfoMessage`. The response will be a collection of `LoginInfoResponseMessage` objects.
```

Not sure if we should add some more context here about the SLNet messages (or even mention that they are SLNet messages...)

```csharp
    public GQIPage GetNextPage(GetNextPageInputArgs args)
    {
        var request = new GetInfoMessage(InfoType.ClientList);
        var responses = _dms.SendMessages(request).OfType<LoginInfoResponseMessage>();
        return default;
    }
```

> [!IMPORTANT]
> DMS messages are subject to change without notice. If you can implement an alternative using the [built-in data sources](xref:Query_data_sources), we highly recommend that you do so instead.

> ## Step 3: Parse the responses into a GQIPage

Reviewer: "Parse" implies interpreting a string.
```suggestion
## Step 3: Transform DMS responses into a GQIPage
```

> 1. Parse the responses into a [GQIPage](xref:GQI_GQIPage).

Reviewer: Idem. Also having a numbered list with only 1 item doesn't really make sense.
Could split this up in smaller steps.
